src/V3ThreadPool.h

```diff
@@ -0,0 +1,192 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Thread pool for Verilator itself
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2005-2023 by Wilson Snyder.  This program is free software; you can
+// redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+
+#ifndef _V3THREADPOOL_H_
+#define _V3THREADPOOL_H_ 1
+
+#include "verilated_threads.h"
+
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <list>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+//============================================================================
+
+class V3ThreadPool final {
+    // MEMBERS
+    using job_t = std::function<void()>;
+
+    mutable VerilatedMutex m_mutex;  // Mutex for use by m_queue and m_cv
+    std::queue<job_t> m_queue VL_GUARDED_BY(m_mutex);  // Queue of jobs
+    std::condition_variable_any m_cv VL_GUARDED_BY(m_mutex);  // Conditions to wake up workers
+    std::list<std::thread> m_workers;  // Worker threads
+    VerilatedMutex m_stoppedJobsMutex;  // Used to signal stopped jobs
+    // Conditions to wake up stopped jobs
+    std::condition_variable_any m_stoppedJobsCV VL_GUARDED_BY(m_stoppedJobsMutex);
+    std::atomic_uint m_stoppedJobs{0};  // Currently stopped jobs waiting for wake up
+
+    bool m_shutdown = false;  // Termination pending
+
+    // CONSTRUCTORS
+    V3ThreadPool() = default;
+    ~V3ThreadPool() {
+        VerilatedLockGuard lock{m_mutex};
+        m_queue = {};  // make sure queue is empty
+        lock.unlock();
+        resize(0);
+    }
+
+public:
+    // METHODS
+    // Singleton
+    static V3ThreadPool& s() VL_MT_SAFE {
+        static V3ThreadPool s_s;
+        return s_s;
+    }
+
+    // Resize thread pool to n workers (queue must be empty)
+    void resize(unsigned n) VL_MT_UNSAFE;
+
+    // Enqueue a job for asynchronous execution
+    template <typename T>
+    std::future<T> enqueue(std::function<T()>&& f) VL_MT_SAFE;
+
+    // Request exclusive access to processing.
+    // It sends request to stop all other threads and waits for them to stop.
+    // Other threads needs to manually call 'check_stop_requested' in places where
+    // they can be stopped.
+    // When all other threads are stopped, this function executes the job
+    // and resumes execution of other jobs.
+    void request_exclusive_access(job_t exclusive_access_job) VL_MT_SAFE {
```

Comment:
```suggestion
    void requestExclusiveAccess(job_t exclusiveAccessJob) VL_MT_SAFE {
```

Verilator style naming please, similar below.